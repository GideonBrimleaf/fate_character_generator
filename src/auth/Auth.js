import React, { useEffect, useState } from 'react';
import { auth } from '../lib/fire';

export const AuthContext = React.createContext();

export const AuthProvider = ({ children }) => {
  const [currentUser, setCurrentUser] = useState(null);
  const [pending, setPending] = useState(true);

  useEffect(() => {
    // TODO - this observer is firing and triggering state
    // changes in test renders which can't be handled with "act",
    // this is creating console warnings in test runs
    auth.onAuthStateChanged((user) => {
      setCurrentUser(user)
      setPending(false)
    });
  }, []);

  if (pending) {
    return <>Loading...</>
  }

  return (
    <AuthContext.Provider
      value={{
        currentUser,
      }}
    >
      {children}
    </AuthContext.Provider>
  );
};
