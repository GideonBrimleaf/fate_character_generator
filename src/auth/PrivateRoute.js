import React, { useContext } from "react";
import { Route, Redirect } from "react-router-dom";
import { AuthContext } from "./Auth";

const PrivateRoute = ({ component: RouteComponent, data, ...rest }) => {
  const {currentUser} = useContext(AuthContext);
  return (
    <Route
      {...rest}
      render={routeProps =>
        !!currentUser ? (
          <RouteComponent {...routeProps} {...data} />
        ) : (
          <Redirect to={"/login"} />
        )
      }
    />
  );
};


export default PrivateRoute