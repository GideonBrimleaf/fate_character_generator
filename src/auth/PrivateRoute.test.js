import React from 'react';
import { render } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';

import { AuthContext } from './Auth';
import PrivateRoute from './PrivateRoute';

it('renders nothing when the user is not logged in', () => {
  const result = render(
    <AuthContext.Provider value={{ currentUser: null }}>
      <MemoryRouter>
        <PrivateRoute />
      </MemoryRouter>
    </AuthContext.Provider>,
  );

  expect(result.asFragment()).toMatchSnapshot();
});

it('renders children when the user is not set', () => {
  const result = render(
    <AuthContext.Provider value={{ currentUser: true }}>
      <MemoryRouter>
        <PrivateRoute component={() => <div id="expected" />} />
      </MemoryRouter>
    </AuthContext.Provider>,
  );

  expect(result.asFragment()).toMatchSnapshot();
});
