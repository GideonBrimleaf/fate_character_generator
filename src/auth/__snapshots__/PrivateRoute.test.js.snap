// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders children when the user is not set 1`] = `
<DocumentFragment>
  <div
    id="expected"
  />
</DocumentFragment>
`;

exports[`renders nothing when the user is not logged in 1`] = `<DocumentFragment />`;
